prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5467,22 +5476,186 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place, encode_kind)?;
 
         Ok((encoded_expr, encoding_stmts, ty, variant_idx))
     }
 
+    fn encode_array_lookup_shared(
+        &mut self,
+        base: PlaceEncoding,
+        index: vir::Expr,
+        lookup_pure_name: String,
+        lookup_pure_ret: vir::Type,
+        array_elem_ty: vir::Type,
+        array_len: usize,
+        val_field: vir::Field,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+
+        let array_type = base.get_type().clone();
+        let lookup_res: vir::Expr = self.cfg_method.add_fresh_local_var(array_elem_ty).into();
+        let lookup_res_val_field = lookup_res.clone().field(val_field.clone());
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Shared)?;
+        stmts.extend(self.encode_havoc_and_allocation(&lookup_res));
+        let idx_val_int = index.field(vir::Field::new("val_int", vir::Type::Int));
+
+        let lookup_pure_call = vir::Expr::func_app(
+            lookup_pure_name,
+            vec![encoded_base_expr.clone(), idx_val_int.clone()],
+            vec![
+                vir::LocalVar::new(
+                    String::from("self"),
+                    array_type.clone(),
+                ),
+                vir::LocalVar::new(
+                    String::from("idx"),
+                    vir::Type::Int,
+                ),
+            ],
+            lookup_pure_ret,
+            vir::Position::default(),
+            );
+
+        stmts.push(vir::Stmt::Assert(
+            vir!{ [idx_val_int.clone()] < [ vir::Expr::from(array_len) ] },
+            vir::Position::default(),
+        ));
+        stmts.push(vir::Stmt::Assert(
+            vir::Expr::predicate_access_predicate(array_type, encoded_base_expr, vir::PermAmount::Read),
+            vir::Position::default(),
+        ));
+
+        stmts.push(vir::Stmt::Inhale(vir!{ [ lookup_pure_call ] == [ lookup_res_val_field ] }));
+        Ok((lookup_res, stmts))
+    }
+
+    fn encode_array_lookup_mut(
+        &mut self,
+        base: PlaceEncoding,
+        index: vir::Expr,
+        lookup_pure_name: String,
+        lookup_pure_ret: vir::Type,
+        array_elem_ty: vir::Type,
+        array_len: usize,
+        val_field: vir::Field,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+        let array_type = base.get_type().clone();
+        let array_type_name = if let vir::Type::TypedRef(ref n) = array_type { n.clone() } else { unreachable!() };
+        let res: vir::Expr = self.cfg_method.add_fresh_local_var(array_elem_ty.clone()).into();
+        let res_val_field = res.clone().field(val_field.clone());
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Mutable)?;
+        let idx_val_int = index.clone().field(vir::Field::new("val_int", vir::Type::Int));
```

Comment:
`Encoder::encode_value_field` is slightly better than hard-coding `"val_int"`.

Actually, now that I think better at it, it's more serious: this might cause a consistency error if nothing else in Prusti calls `encode_value_field` to generate a `val_int` field. It's rare, but if it happens then the Viper program would not have the`field val_int: Int` declaration that is generated by `Encoder::encode_value_field`.

Reply:
agreed, using [`TyCtxt::mk_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.mk_ty) now and then encoding the freshly-created `usize` type for the `index field`.